prow/owners/owners.go

```diff
@@ -0,0 +1,477 @@
+		return nil
+	}
+
+	if err := yaml.Unmarshal(b, c); err != nil {
+		log.WithError(err).Errorf("Failed to unmarshal file contents.")
+		return nil
+	}
+
+	relPath, err := filepath.Rel(o.baseDir, path)
+	if err != nil {
+		log.WithError(err).Errorf("Unable to find relative path between baseDir: %q and path.", o.baseDir)
+		return err
+	}
+	relPathDir := canonicalize(filepath.Dir(relPath))
+	o.applyConfigToPath(relPathDir, c)
+	return nil
+}
+
+// decodeOwnersMdConfig will parse the yaml header if it exists and unmarshal it into an ownersConfig.
+// If no yaml header is found, do nothing
+// Returns an error if the file cannot be read or the yaml header is found but cannot be unmarshalled
+var mdStructuredHeaderRegex = regexp.MustCompile("^---\n(.|\n)*\n---")
+
+func decodeOwnersMdConfig(path string, config *ownersConfig) error {
+	fileBytes, err := ioutil.ReadFile(path)
+	if err != nil {
+		return err
+	}
+	// Parse the yaml header from the top of the file.  Will return an empty string if regex does not match.
+	meta := mdStructuredHeaderRegex.FindString(string(fileBytes))
+
+	// Unmarshal the yaml header into the config
+	return yaml.Unmarshal([]byte(meta), &config)
+}
+
+func normLogins(logins []string) sets.String {
+	normed := sets.NewString()
+	for _, login := range logins {
+		normed.Insert(github.NormLogin(login))
+	}
+	return normed
+}
+
+func (o *RepoOwners) applyConfigToPath(path string, config *ownersConfig) {
+	if len(config.Approvers) > 0 || len(config.Assignees) > 0 {
+		o.approvers[path] = o.ExpandAliases(normLogins(config.Approvers).Union(normLogins(config.Assignees)))
+	}
+	if len(config.Reviewers) > 0 {
+		o.reviewers[path] = o.ExpandAliases(normLogins(config.Reviewers))
+	}
+	if len(config.Labels) > 0 {
+		o.labels[path] = sets.NewString(config.Labels...)
+	}
+}
+
+func (o *RepoOwners) filterCollaborators(toKeep []github.User) *RepoOwners {
+	collabs := sets.NewString()
+	for _, keeper := range toKeep {
+		collabs.Insert(github.NormLogin(keeper.Login))
+	}
+
+	filter := func(ownerMap map[string]sets.String) map[string]sets.String {
+		filtered := make(map[string]sets.String)
+		for path, unfiltered := range ownerMap {
+			filtered[path] = unfiltered.Intersection(collabs)
+		}
+		return filtered
+	}
+
+	result := *o
+	result.approvers = filter(o.approvers)
+	result.reviewers = filter(o.reviewers)
+	return &result
+}
+
+// findOwnersForPath returns the OWNERS file path furthest down the tree for a specified file
+// By default we use the reviewers section of owners flag but this can be configured by setting approvers to true
+func findOwnersForPath(path string, ownerMap map[string]sets.String) string {
+	d := path
+
+	for {
+		n, ok := ownerMap[d]
+		if ok && len(n) != 0 {
+			return d
+		}
+		if d == baseDirConvention {
+			break
+		}
+		d = filepath.Dir(d)
+		d = canonicalize(d)
+	}
+	return ""
+}
+
+// FindApproversForPath returns the OWNERS file path furthest down the tree for a specified file
```

Comment:
s/FindApproversForPath/FindApproverOwnersForPath/

Reply:
Fixed. Also changed the analogous reviewers function to a more consistent name.